cirq-core/cirq/experiments/two_qubit_xeb.py

```diff
@@ -0,0 +1,223 @@
+# Copyright 2024 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Sequence, TYPE_CHECKING, Optional, Tuple, Dict
+
+from dataclasses import dataclass
+import itertools
+import functools
+
+from matplotlib import pyplot as plt
+import networkx as nx
+import numpy as np
+import pandas as pd
+
+from cirq import ops, devices, value, vis
+from cirq.experiments.xeb_sampling import sample_2q_xeb_circuits
+from cirq.experiments.xeb_fitting import benchmark_2q_xeb_fidelities
+from cirq.experiments.xeb_fitting import fit_exponential_decays, exponential_decay
+from cirq.experiments import random_quantum_circuit_generation as rqcg
+
+if TYPE_CHECKING:
+    import cirq
+
+
+def _grid_qubits_for_sampler(sampler: 'cirq.Sampler'):
+    if hasattr(sampler, 'processor'):
+        device = sampler.processor.get_device()
+        return sorted(device.metadata.qubit_set)
+    else:
+        qubits = devices.GridQubit.rect(3, 2, 4, 3)
+        # Delete one qubit from the rectangular arangement to
+        # 1) make it irregular 2) simplify simulation.
+        return qubits[:-1]
+
+
+def _manhattan_distance(qubit1: 'cirq.GridQubit', qubit2: 'cirq.GridQubit') -> int:
+    return abs(qubit1.row - qubit2.row) + abs(qubit1.col - qubit2.col)
+
+
+@dataclass(frozen=True)
+class TwoQubitXEBResult:
+    """Results from an XEB experiment."""
+
+    fidelities: pd.DataFrame
+
+    @functools.cached_property
+    def _qubit_pair_map(self) -> Dict[Tuple['cirq.GridQubit', 'cirq.GridQubit'], int]:
+        return {
+            (min(q0, q1), max(q0, q1)): i
+            for i, (_, _, (q0, q1)) in enumerate(self.fidelities.index)
+        }
+
+    @functools.cached_property
+    def all_qubit_pairs(self) -> Tuple[Tuple['cirq.GridQubit', 'cirq.GridQubit'], ...]:
+        return tuple(sorted(self._qubit_pair_map.keys()))
+
+    def plot_heatmap(self, ax: Optional[plt.Axes] = None, **plot_kwargs):
```

Comment:
Please return plt.Axes from plot functions, for example like here - https://github.com/quantumlib/Cirq/blob/8e47acd3a2b4a4c30d7f35992a8a001f7a06cc49/cirq-core/cirq/experiments/t2_decay_experiment.py#L396

Reply:
done